core/env.go

```diff
@@ -0,0 +1,655 @@
+	case envconfig.SDKPython:
+		return pythonRuntime(ctx, bk, progSock, pipeline, env.Platform, env.Directory, env.ConfigPath)
+	default:
+		return nil, fmt.Errorf("unknown sdk %q", env.Config.SDK)
+	}
+}
+
+func execEntrypoint(
+	ctx context.Context,
+	bk *buildkit.Client,
+	progSock string,
+	pipeline pipeline.Path,
+	envCache *EnvironmentCache,
+	env *Environment,
+	entrypointName string,
+	args any,
+) ([]byte, error) {
+	ctx, err := envCache.ContextWithCachedEnv(ctx, env)
+	if err != nil {
+		return nil, fmt.Errorf("failed to set environment in context: %w", err)
+	}
+
+	inputMap := map[string]any{
+		// TODO: remember to tell Helder that this is a small breaking change, need to tweak python sdk code.
+		// "resolver" used to be in form <parent>.<field>, now its just the name of the entrypoint (i.e. check
+		// name, artifact name, etc.)
+		"resolver": entrypointName,
+		"args":     args,
+		"parent":   nil, // for now, could support parent data in future for user-defined chainable types
+	}
+	inputBytes, err := json.Marshal(inputMap)
+	if err != nil {
+		return nil, err
+	}
+	ctr, err := env.Runtime.WithNewFile(ctx, bk, filepath.Join(inputMountPath, inputFile), inputBytes, 0644, "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to mount resolver input file: %w", err)
+	}
+
+	ctr, err = ctr.WithMountedDirectory(ctx, bk, outputMountPath, NewScratchDirectory(pipeline, ctr.Platform), "")
+	if err != nil {
+		return nil, fmt.Errorf("failed to mount resolver output directory: %w", err)
+	}
+
+	ctr, err = ctr.WithExec(ctx, bk, progSock, ctr.Platform, ContainerExecOpts{
+		ExperimentalPrivilegedNesting: true,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to exec resolver: %w", err)
+	}
+	err = ctr.Evaluate(ctx, bk)
+	if err != nil {
+		return nil, fmt.Errorf("failed to evaluate resolver: %w", err)
+	}
+
+	outputFile, err := ctr.File(ctx, bk, filepath.Join(outputMountPath, outputFile))
+	if err != nil {
+		return nil, fmt.Errorf("failed to get resolver output file: %w", err)
+	}
+
+	entrypointOutput, err := outputFile.Contents(ctx, bk)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read resolver output file: %w", err)
+	}
+
+	return entrypointOutput, nil
+}
+
+func (env *Environment) WithCheck(check *Check, envCache *EnvironmentCache) (*Environment, error) {
+	env = env.Clone()
+	env.Checks = append(env.Checks, check)
+	envID, err := env.ID()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get environment ID: %w", err)
+	}
+	if err := env.setEntrypointEnvs(envID); err != nil {
+		return nil, fmt.Errorf("failed to set entrypoint envs: %w", err)
+	}
+	return env, nil
+}
+
+type CheckID string
+
+func (id CheckID) String() string {
+	return string(id)
+}
+
+func (id CheckID) ToCheck() (*Check, error) {
+	var check Check
+	if id == "" {
+		return &check, nil
+	}
+	if err := resourceid.Decode(&check, id); err != nil {
+		return nil, err
+	}
+	if !check.IsCheck {
```

Comment:
Could be worth cherry-picking this if you agree with the idea: https://github.com/sipsma/dagger/commit/45fc541290ae0d3609a5c6d3f82718f805e222f7

It changes IDs to look like `core.ContainerID:eyJ...` and verifies the type matches the expected type when decoding. It's a little janky that it's tightly coupled to Go's `%T` representation, so that could be tweaked, but this is all internal plumbing at the moment anyway.

Tangentially related, I came across [RFC-6920](https://www.rfc-editor.org/rfc/rfc6920) today (via [this post](https://posts.summerti.me/rfc-6920/)) which seems like a reasonable candidate format for our IDs. Maybe something like this?:

```
ni://dagger.io/sha-256;f4OxZX_x_FO5LcGBSKHWXfwtSx-j1ncoSt3SABJtkGk?ct=application/x-dagger-container
```

Anyway, that's neither here nor there, just cool to see a relevant RFC, with some kind of fun implications like having `https://dagger.io/.well-known/ni/sha256/f4OxZX_x_FO5LcGBSKHWXfwtSx-j1ncoSt3SABJtkGk` do... something interesting.

Reply:
Oh nice thank you! I have a `TODO:` comment somewhere about backporting the cleaner fix to all this but hadn't taken the time to go track down which commit it was, appreciate it!

> Tangentially related, I came across [RFC-6920](https://www.rfc-editor.org/rfc/rfc6920) today (via [this post](https://posts.summerti.me/rfc-6920/)) which seems like a reasonable candidate format for our IDs. Maybe something like this?:

Ah neat! Honestly I'm still holding out that someday we'll have time to finally do the work needed to persist hash<->id mapping in the cache and be rid of needing to deal with anything like this, *but* the URL approach is super intriguing nonetheless since it would still probably make sense to support something akin to the current IDs externally as a "serialized DAG" format (similar to what you did w/ Bass IIRC)

Reply:
> Tangentially related, I came across [RFC-6920](https://www.rfc-editor.org/rfc/rfc6920) today (via [this post](https://posts.summerti.me/rfc-6920/)) which seems like a reasonable candidate format for our IDs. Maybe something like this?:

I love that too

Quick question tho about the article:

> And while [CDNs do support caching subsets of a whole object](https://docs.aws.amazon.com/AmazonCloudFront/latest/DeveloperGuide/RangeGETs.html), there is no easy way for a RFC 6920 client to check whether their partial content is valid or not, as it uses a hash over the whole data, and not over a tree of parts.

Can this create issue with our cache? Like valid cache that should be invalidated or vice-versa? I do not have the full context of the id work but I'm trying to understand the whole logic.

Reply:
> Can this create issue with our cache? Like valid cache that should be invalidated or vice-versa? I do not have the full context of the id work but I'm trying to understand the whole logic.

@TomChv I don't think we would need to care about partial fetches/etc, since the object being referenced is relatively small - essentially a recipe for building a container image, and not the container image itself.